HIP/hip-655.md

```diff
@@ -0,0 +1,106 @@
+author: Nick Poorman <@nickpoorman>
+working-group: Richard Bair <@rbair23>, Jasper Potts <@jasperpotts>, Atul Mahamuni <@atul-hedera>, Michael Tinker <@tinker-michaelj>, Leemon Baird <@lbaird>
+type: Standards Track 
+category: Service
+needs-council-approval: Yes 
+status: Review
+created: 2022-10-20 
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/654
+updated: 2023-01-20, 2023-01-24
+---
+
+## Abstract
+
+[HIP-23](https://hips.hedera.com/hip/hip-23) allowed an account owner to open up a specified number of automatic token-association slots into which others can transfer tokens. However, this capability doesn't allow for a pure "airdrop" of tokens. This HIP provides an additional mechanism for senders to airdrop tokens to other accounts, while allowing individual account holders to maintain control over what shows up in their account balances.
+
+## Motivation
+
+The current implementation of the Hedera Token Service does not allow airdrops of arbitrary tokens to any account. [HIP-23](https://hips.hedera.com/hip/hip-23) lists several disadvantages of arbitrary airdrops - such as exposing an innocent account holder to reputational or legal or spam risk of being associated with unwanted/unsolicited tokens, as well as DDoS risks for the network. And it proposes opt-in token associations. Further, [HIP-367](https://hips.hedera.com/hip/hip-367) removes the limit on token associations. Even with these enhancements, users still cannot airdrop tokens using the Hedera network today.
+
+## Rationale
+
+Token airdrops are very common in the blockchain industry and many legitimate use cases require airdropping of tokens. This HIP attempts to provide a mechanism to enable these airdrops while still protecting the account owners, and imposing dis-incentivizing costs to the senders of the airdropped tokens.
+
+## User Stories
+
+### Terminology
+
+The following terminology is used in the rest of this Specification
+
+- Airdrop-Sender: the owner of a token that they wish to send to somebody without the explicit association/permission from the intended recipient
+- Airdrop-Recipient: the intended recipient of the above airdrop
+- Node-Operator: the operator of a node in the Hedera network
+    
+### User stories
+    
+- As an Airdrop-Sender, I want to send arbitrary tokens (Fungible and/or Non-fungible) to Airdrop-Recipients so that I can create interesting business use cases
+    1. I understand that the recipients of the airdrop may not wish to accept these tokens and I am willing to pay for the network resources used while they make their decision.
+    2. If the intended Airdrop-Recipients do not accept my transfer, I want to be able to restore those tokens to my account so that I can do whatever I wish with those tokens.
+- As an Airdrop-Recipient, I want to receive arbitrary tokens (Fungible and/or Non-fungible) from the Airdrop-Senders so that I can accumulate those in my account.
```

Comment:
4. I want to reject an airdrop and remove it from my inbox without paying any fees to the network.

Reply:
Ignoring a scheduled transaction will accomplish this.

Reply:
The sender's wallet queries for association, if not found creates a scheduled transaction, the receiver's wallet queries a mirror node for inbound scheduled transactions, recognizes air-drops, prompts user for confirmation, makes the association, then signs the scheduled transaction which is set to execute on sufficient signatures.

OR

sender marks their NFT as an allowance, receiver creates and signs a TX moving the NFT to their own wallet

The UI experience can be streamlined for one-touch approval on the receiver's part.  If they don't want to be bothered, that is what auto-association is for.

I still don't see why this HIP is necessary, particularly given the amount of engineering required.

Reply:
As a User, I do not want to be bothered with yet another Inbox that is full of spam, or any feature that encourages wallet developers to display a bright red badge showing the 1000's of random solicitations I have not yet acknowledged, accepted, or moved to junk.

Reply:
Thanks for the further clarification on how you're suggesting this functionality without the HIP.

I can't understand how this would actually work in production. The linear pattern of calls makes sense but there's some assumptions and a lot of querying the network required by multiple parties to complete an airdrop.

Suggested flow without needing this HIP:
> 1. The sender's wallet queries for association, if not found creates a scheduled transaction
This makes sense to me, already how most execute airdrops.

> 2. The receiver's wallet queries a mirror node for inbound scheduled transactions, recognizes air-drops, prompts user for confirmation

This I don't get think is a scalable solution at all for developers. 

Feels like every wallet id has to continually check the mirror node to see if there's any transaction bytes trying to be sent to them. 

That's a lot of work if say Hashpack has to continually query for all wallets / or all transaction bytes to alert / show users that they've received an event. 

To me, this is why we need the HIP. You shouldn't have to query and build large scale solutions externally to the network to know when a very common event has happened (a user gets airdropped a token)

> 3. User makes the association, then signs the scheduled transaction which is set to execute on sufficient signatures.

This again requires a lot of work for wallet providers to not only continually monitor transaction bytes, realize it's an association, wait for user confirmation, then get the CORRECT transaction bytes that is the airdrop that correspond with that same token id. (1 account can send multiple transactions of multiple tokens, not sure how you'd be able to match bytes to token Id in that case honestly)


**The other method suggested:**
> Sender marks their NFT as an allowance, receiver creates and signs a TX moving the NFT to their own wallet

This requires a **user** to create a transaction to accept the token allowance  and then accept the NFT (as I understand how allowance works). 

**That's sort of the exact opposite from what we're looking for and why I believe this HIP is needed:**
No action needed from user to be able to accept an airdrop, while maintaining security to not accept all sent.

Added feature I very much agree with suggested by @rocketmay, allowing sender airdrop account to pay for association of that token to the account to allow empty wallets to receive earned FTs or NFTs.

Reply:
> As a User, I do not want to be bothered with yet another Inbox that is full of spam, or any feature that encourages wallet developers to display a bright red badge showing the 1000's of random solicitations I have not yet acknowledged, accepted, or moved to junk.

I agree that this should be a state of the inbox. Reject all tokens sent to me that aren't associated. Replication of how this currently works makes a lot of sense to me. 

I added this use case here:
https://github.com/hashgraph/hedera-improvement-proposal/pull/655/files#r1095099366

Reply:
> query for inbound scheduled transactions,

I'm not aware of this capability. Can you post a link to documentation around this?

Reply:
> 4. I want to reject an airdrop and remove it from my inbox without paying any fees to the network.

I believe a rejection would remove the record stored in your `unsolicited-tokens-inbox`. A one sided record, where the sender has a record in their `unsolicited-tokens-outbox`, and the receiver no longer has a record in their `unsolicited-tokens-inbox` would signify a rejection. I'm not entirely sure if that's how we would do it in practice. If that's the case the sender would need the ability to remove the record from their `unsolicited-tokens-outbox`. If that's not the case, I would need to think through more on how we could derive a rejected record other than stitching together historical data.